File: src/web/cocos2d/src/zotohlab/ext/negamax.js
// This library is distributed in  the hope that it will be useful but without
// any  warranty; without  even  the  implied  warranty of  merchantability or
// fitness for a particular purpose.
// The use and distribution terms for this software are covered by the Eclipse
// Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
// can be found in the file epl-v10.html at the root of this distribution.
// By using this software in any  fashion, you are agreeing to be bound by the
// terms of this license. You  must not remove this notice, or any other, from
// this software.
// Copyright (c) 2013-2015, Ken Leung. All rights reserved.

/**
 * @requires cherimoia/skarojs
 * @module zotohlab/asx/negamax
 */
define("zotohlab/asx/negamax",

  ['cherimoia/skarojs'],

  function (sjs) { "use strict";

    const PINF = 1000000;

    /** @alias module:zotohlab/asx/negamax */
    let exports = {},
    undef;

    //////////////////////////////////////////////////////////////////////////////
    let negamax = (board, game, maxDepth, depth, alpha, beta) => {
      if (depth === 0 || board.isOver(game)) {
        return board.evalScore(game);
      }
      //assert(openMoves && openMoves.length > 0);
      let openMoves = board.getNextMoves(game),
      bestValue = -PINF,
      rc,
      n,
      move,
      bestMove = openMoves[0];

      if (depth === maxDepth) {
        game.lastBestMove = openMoves[0];
      }

      for (n=0; n < openMoves.length; ++n) {
        move = openMoves[n];
        board.makeMove(game, move);
        board.switchPlayer(game);
        rc = - negamax(board, game, maxDepth, depth-1, -beta, -alpha);
        board.switchPlayer(game);
        board.unmakeMove(game,move);
        bestValue = Math.max( bestValue,  rc );
        if (alpha < rc) {
          alpha = rc;
          bestMove = move;
          if (depth === maxDepth) {
            game.lastBestMove = move;
          }
          if (alpha >= beta) { break; }
        }
      };
      return bestValue;
    }

    /**
     * @interface GameBoard
     * @class
     */
    class GameBoard {
      /**
       * @memberof module:zotohlab/asx/negamax~GameBoard
       * @method isOver
       * @param {Object} game
       * @return {Boolean}
       */
      isOver(game) { return false; }
      /**
       * @memberof module:zotohlab/asx/negamax~GameBoard
       * @method evalScore
       * @param {Object} game
       * @return {Number} score
       */
      evalScore(game) { return 0; }
      /**
       * @memberof module:zotohlab/asx/negamax~GameBoard
       * @method getNextMoves
       * @param {Object} game
       * @return {Array}
       */
      getNextMoves(game) { return []; }
      /**
       * @memberof module:zotohlab/asx/negamax~GameBoard
       * @method makeMoves
       * @param {Object} game
       * @param {Number} move
       */
      makeMove(game,move) {}
      /**
       * @memberof module:zotohlab/asx/negamax~GameBoard
       * @method switchPlayer
       * @param {Object} game
       */
      switchPlayer(game) {}
      /**
       * @memberof module:zotohlab/asx/negamax~GameBoard
       * @method unmakeMove
       * @param {Object} game
       * @param {Number} move
       */
      unmakeMove(game,move) {}
      /**
       * @memberof module:zotohlab/asx/negamax~GameBoard
       * @method takeSnapshot
       * @return {SnapShot}
       */
      takeSnapshot() { return null; }
    }

    /**
     * The negamax algorithm implementation.
     * @class NegaMax
     */
    class NegaMax {

      /**
       * Get the game board.
       * @memberof module:zotohlab/asx/negamax~NegaMax
       * @method getGameBoard
       * @return {GameBoard}
       */
      getGameBoard() { return this.board; }

      /**
       * Constructor.
       *
       * @memberof module:zotohlab/asx/negamax~NegaMax
       * @method constructor
       * @param {GameBoard} board
       */
      constructor(board) { this.board= board; }

      /**
       * Run the algo for one iteration.
       *
       * @memberof module:zotohlab/asx/negamax~NegaMax
       * @method eval
       * @return {Number} last best move
       */
      eval() {
        const snapshot= this.board.takeSnapshot();
        negamax(this.board, snapshot, 10, 10, -PINF, PINF);
        return snapshot.lastBestMove;
      }

    }

    /**
     * Simple data structure keeping track of the state of the
     * game board.
     * @class Snapshot
     */
    class Snapshot {

      /**
       * @method ctor
       * @private
       */
      constructor() {
        this.lastBestMove= null;
        this.other= null;
        this.cur= null;
        this.state= null;
      }

    }

    exports = /** @lends exports# */{
      /**
       * @property {SnapShot} Snapshot
       */
      Snapshot: Snapshot,
      /**
       * @property {Number} INF
       */
      INF: PINF,
      /**
       * @property {NegaMax} NegaMax
       */
      NegaMax: NegaMax,
      /**
       * @property {GameBoard} GameBoard
       */
      GameBoard: GameBoard
    };

    return exports;
});

//////////////////////////////////////////////////////////////////////////////
//EOF

File: src/main/clojure/czlab/frigga/core/negamax.clj
;; This library is distributed in  the hope that it will be useful but without
;; any  warranty; without  even  the  implied  warranty of  merchantability or
;; fitness for a particular purpose.
;; The use and distribution terms for this software are covered by the Eclipse
;; Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
;; can be found in the file epl-v10.html at the root of this distribution.
;; By using this software in any  fashion, you are agreeing to be bound by the
;; terms of this license. You  must not remove this notice, or any other, from
;; this software.
;; Copyright (c) 2013-2015, Ken Leung. All rights reserved.

(ns ^{:doc ""
      :author "kenl"}

  czlab.frigga.core.negamax

  (:require
    [czlab.xlib.util.core :refer [MubleObj! ]]
    [czlab.xlib.util.logging :as log]
    [czlab.xlib.util.str :refer [strim hgl?]])

  (:import
    [com.zotohlab.odin.game Game PlayRoom
    Board Player PlayerSession]
    [com.zotohlab.skaro.core Muble]
    [com.zotohlab.odin.core Session]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;(set! *warn-on-reflection* true)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(def ^:private PINF 1000000)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defprotocol NegaAlgoAPI "" (Evaluate [_]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defprotocol NegaSnapshotAPI

  ""

  (SetLastBestMove [_ m] )
  (SetOther [_ o] )
  (SetCur [_ c] )
  (SetState [_ s] )
  (LastBestMove [_] )
  (Other [_] )
  (Cur [_] )
  (State [_] ))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn NegaMax

  "The Nega-Max algo implementation"

  [^Board board game
   maxDepth depth alpha beta]

  (if (or (== depth 0)
          (.isOver board game))
    (.evalScore board game)
    ;;:else
    (with-local-vars
      [openMoves (.getNextMoves board game)
       bestValue (- PINF)
       localAlpha alpha
       halt false
       rc 0
       bestMove (nth openMoves 0) ]
      (when (== depth maxDepth)
        (SetLastBestMove game (nth @openMoves 0))) ;; this will change overtime, most likely
      (loop [n 0]
        (when-not (or (> n (count @openMoves))
                      (true? @halt))
          (let [move (nth @openMoves n) ]
            (doto board
              (.makeMove game move)
              (.switchPlayer game))
            (var-set rc (- (NegaMax board
                                    game
                                    maxDepth
                                    (dec depth)
                                    (- beta) (- @localAlpha))))
            (doto board
              (.switchPlayer game)
              (.unmakeMove game move))
            (var-set bestValue (Math/max (long @bestValue) (long @rc)))
            (when (< @localAlpha @rc)
              (var-set localAlpha @rc)
              (var-set bestMove move)
              (when (== depth maxDepth)
                (SetLastBestMove game move))
              (when (>= @localAlpha beta)
                (var-set halt true)))
            (recur (inc n) ))))
      @bestValue)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn ReifyNegaMaxAlgo

  "Create an implementation of nega-max"

  [^Board board]

  (reify
    NegaAlgoAPI
    (Evaluate [_]
      (let [snapshot (.takeSnapshot board) ]
        (NegaMax board snapshot 10 10 (- PINF) PINF)
        (LastBestMove snapshot)))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn ReifySnapshot

  "Create a snapshot"

  []

  (let [impl (MubleObj!)]
    (reify
      NegaSnapshotAPI
      (SetLastBestMove [_ m] (.setv impl :lastbestmove m))
      (SetOther [_ o] (.setv impl :other o))
      (SetCur [_ c] (.setv impl :cur c))
      (SetState [_ s] (.setv impl :state s))
      (LastBestMove [_] (.getv impl :lastbestmove))
      (Other [_] (.getv impl :other))
      (Cur [_] (.getv impl :cur))
      (State [_] (.getv impl :state)))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;EOF

